### DISCUSSION TITLE
XML Parser should support OWL/XML

### DISCUSSION THREAD

**Discussion by jpmccu** [RDFLib/rdflib#1571] [Ideas]
The detection should be when the OWL namespace is declared instead of the RDF namespace. In fact, the namespace should be what triggers RDF as well, since RDF can be embedded in other non RDF/XML documents as well.

On a practical note, it can be difficult to tell if someone has published RDF/XML vs OWL/XML from just the file extension. This should successfully parse, after this is implemented:

```python
import rdflib

ont_url = 'https://raw.githubusercontent.com/ISA-tools/stato/dev/releases/1.2/stato.owl'
file_format = rdflib.util.guess_format(ont_url)
g = rdflib.Graph()
g.parse(location=ont_url,format=file_format)
```

Currently, we get:

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/apps/satoru/venv/local/lib/python2.7/site-packages/rdflib/graph.py", line 1032, in parse
    parser.parse(source, self, **args)
  File "/apps/satoru/venv/local/lib/python2.7/site-packages/rdflib/plugins/parsers/rdfxml.py", line 579, in parse
    self._parser.parse(source)
  File "/usr/lib/python2.7/xml/sax/expatreader.py", line 107, in parse
    xmlreader.IncrementalParser.parse(self, source)
  File "/usr/lib/python2.7/xml/sax/xmlreader.py", line 123, in parse
    self.feed(buffer)
  File "/usr/lib/python2.7/xml/sax/expatreader.py", line 210, in feed
    self._parser.Parse(data, isFinal)
  File "/usr/lib/python2.7/xml/sax/expatreader.py", line 352, in end_element_ns
    self._cont_handler.endElementNS(pair, None)
  File "/apps/satoru/venv/local/lib/python2.7/site-packages/rdflib/plugins/parsers/rdfxml.py", line 161, in endElementNS
    self.current.end(name, qname)
  File "/apps/satoru/venv/local/lib/python2.7/site-packages/rdflib/plugins/parsers/rdfxml.py", line 333, in node_element_end
    self.error("Repeat node-elements inside property elements: %s"%"".join(name))
  File "/apps/satoru/venv/local/lib/python2.7/site-packages/rdflib/plugins/parsers/rdfxml.py", line 187, in error
    raise ParserError(info + message)
rdflib.exceptions.ParserError: http://purl.obolibrary.org/obo/stato.owl:17:117: Repeat node-elements inside property elements: http://www.w3.org/2002/07/owl#Literal
```

**Comment by mwatts15** [comment]
@jimmccusker what exactly do you mean by "when the OWL namespace is declared instead of the RDF namespace"? Currently, the document you linked has both `xmlns:owl` and `xmlns:rdf` declared in the root element.

**Comment by mwatts15** [comment]
Also worth noting, the linked document is currently parsed without error by the RDF/XML parser with the current HEAD of the master branch. We don't have a distinct OWL format.

**Comment by jpmccu** [comment]
It looks like this particular ontology transitioned over to rdf/xml. Here is the one that would have been loaded when I filed the bug:

https://raw.githubusercontent.com/ISA-tools/stato/dev/releases/1.2/stato.owl

I've updated the issue to use the older version.

**Comment by ghost** [comment]
The original of that ontology is ill-formed, several Literals contain unescaped double-quotes. There's a commented-out fix in 
Jean-Baptiste Lamy's [owlready2 owl_2_ntriples implementation](https://bitbucket.org/jibalamy/owlready2/src/26369f19626aaf0a10243d8c1489f50c9eb62629/owlxml_2_ntriples.py#lines-145) which I uncommented and with a bit of quick-and-dirty hacking I adapted his parser for use as an RDFLib plugin-parser for OWL/XML - which _does_ read the original ontology (ill-formed as it is) into an RDFLib Graph. Pro tem, I have it earmarked for the RDFLib Cookbook.

**Reply by tommycarstensen** [reply]
@gjhiggins Can you post your solution here? I tried the following:

```from owlready2 import get_ontology
from owlready2 import default_world
onto = get_ontology(uri)
g = default_world.as_rdflib_graph()
g.serialize(format='turtle')
```


I also tried:
```from owlready2 import owlxml_2_ntriples
nb_triple = owlxml_2_ntriples.parse(uri)
```


But the latter gave me this error:
`owlready2.base.OwlReadyOntologyParsingError: OWL/XML parsing error in file bto.owl, line 31, column 83.`


This is the input file:
`https://raw.githubusercontent.com/BRENDA-Enzymes/BTO/master/bto.owl`

**Reply by ghost** [reply]
> @gjhiggins Can you post your solution here?

Sure. But you don't actually *need* it. Thank you for posting the URL to the input, that makes it easier to respond positively -- happily, the standard RDFLib XML parser successfully parses your BTO ontology:

```python
import rdflib

def test_bto():
    g = rdflib.Graph()
    g.parse(
        location="https://raw.githubusercontent.com/BRENDA-Enzymes/BTO/master/bto.owl",
        format="xml",
    )
    assert len(g) == 87688
```
But I'll post the solution here, just in case someone finds it useful. ***Caveat*** YMMV - it successfully parses the [STATO](http://purl.obolibrary.org/obo/stato.owl) ontology but not the [UBERON](http://purl.obolibrary.org/obo/uberon.owl) ontology (and I haven't yet investigated why, however, the standard RDFLib XML parser successfully parses it).

Assuming the unzipped attachment  [owlxml.py.gz](https://github.com/RDFLib/rdflib/files/9800071/owlxml.py.gz) is copied into `rdflib.plugins.parsers` as `owlxml.py`, then including the relevant plugin declaration will enable it ...

I tested it with a local copy of `stato.owl` in `test/owl`.
```python
import rdflib
from test.data import TEST_DATA_DIR

rdflib.plugin.register(
    "owl",  # Use this string as arg to “format”
    rdflib.parser.Parser,
    "rdflib.plugins.parsers.owlxml",
    "OWLXMLParser",
)

testloc = TEST_DATA_DIR / "owl" / "stato.owl"


def test_owlxmlparser():
    g = rdflib.Graph()
    g.parse(
        location=testloc,
        format="owl",  # Using registered format-selecting string
    )
    assert len(g) == 13769

```

HTH

**Comment by malinjiangcs** [comment]
Standard RDFLib still cannot successfully parse owl/xml. Will this function be added？

**Reply by aucampia** [reply]
@malinjiangcs if you submit a PR we will have a look at it, but there is no ongoing work on supporting it and I dont think we plan to add it. We support RDF/XML and you can use that to encode OWL.

https://www.w3.org/TR/owl2-primer/#OWL_Syntaxes
> There are various syntaxes available for OWL which serve various purposes. The Functional-Style syntax [[OWL 2 Specification](https://www.w3.org/TR/owl2-primer/#ref-owl-2-specification)] is designed to be easier for specification purposes and to provide a foundation for the implementation of OWL 2 tools such as APIs and reasoners. The RDF/XML syntax for OWL is just RDF/XML, with a particular translation for the OWL constructs [[OWL 2 RDF Mapping](https://www.w3.org/TR/owl2-primer/#ref-owl-2-rdf-mapping)]. This is the only syntax that is mandatory to be supported by all OWL 2 tools. The Manchester syntax [[OWL 2 Manchester Syntax](https://www.w3.org/TR/owl2-primer/#ref-owl-2-manchester-syntax)] is an OWL syntax that is designed to be easier for non-logicians to read. The OWL XML syntax is an XML syntax for OWL defined by an XML schema [[OWL 2 XML](https://www.w3.org/TR/owl2-primer/#ref-owl-2-xml-serialization)]. There are tools that can translate between the different syntaxes for OWL. In many syntactic forms, OWL language constructs are also represented by IRIs, and some declarations might be needed to use the abbreviated forms as in the examples. Again, necessary details are found in [Section 8.2.](https://www.w3.org/TR/owl2-primer/#Ontology_Management)
